doc/developer-guide/internal-libraries/MemArena.en.rst

```diff
@@ -14,223 +14,140 @@
-
-.. code-block:: cpp
-   
-   ts::MemArena *arena = new ts::MemArena(64);
-
-The caller can actually allocate more than 64 bytes. 
-
-.. code-block:: cpp
-
-   ts::MemSpan span1 = arena->alloc(16);
-   ts::MemSpan span1 = arena->alloc(256);
-
-Now, span1 and span2 are in the same generation and can both be safely used. After:
-
-.. code-block:: cpp
-
-   arena->freeze();
-   ts::MemSpan span3 = arena->alloc(512);
-   arena->thaw();
-
-span3 can still be used but span1 and span2 have been deallocated and usage is undefined. 
-
-Internal blocks are adjusted for optimization. Each :class:`MemArena::Block` is just a header for the underlying memory it manages. The header and memory are allocated together for locality such that each :class:`MemArena::Block` is immediately followed with the memory it manages. If a :class:`MemArena::Block` is larger than a page (defaulted at 4KB), it is aligned to a power of two. The actual memory that a :class:`MemArena::Block` can allocate out is slightly smaller. This is because a portion of the allocated memory is reserved for the header. Another 16 bytes is reserved to track the allocation headers used by malloc; for page alignment. ie, the default block size is 32768 bytes, but it will only be able to allocate out 32720 bytes. 
+When a :class:`MemArena` instance is constructed it can be constructed with either no allocated
+memory or an initial memory block of a specific size. :func:`MemArena::reserve` can be used to
+specify the size of the next internally allocated block if a specific allocation size is needed.
+This allows allocation to be delayed until memory is actually needed (via a call to
+:func:`MemArena::alloc`).
+
+In normal use memory is allocated from :class:`MemArena` using :func:`MemArena::alloc` to get chunks
+of memory, or :func:`MemArena::make` to get constructed class instances. :func:`MemArena::make`
+takes an arbitrary set of arguments which it attempts to pass to a constructor for the type
+:code:`T` after allocating memory (:code:`sizeof(T)` bytes) for the object. If there isn't enough
+free memory, a new internal block is allocated. Each time a new internal block is allocated, the
+block than the previously allocated one. In general this suffices but in certain cases, if it is
+necessary adjust the internal allocation size, he size can be controlled using
+:func:`MemArena::reserve`. Other than a freeze / thaw cycle, there is no mechanism to release memory
+except for the destruction of the :class:`MemArena`. In such use cases either wasted memory must be
+small enough or temporary enough to not be an issue, or there must be a provision for some sort of
+garbage collection.
+
+Generally :class:`MemArena` is not as useful for classes that allocate their own internal memory
+(such as :code:`std::string` or :code:`std::vector`), which includes most container classes. One
+container class that can be easily used is :class:`IntrusiveDList` because the links are in the
```

Comment:
`IntrusiveDList` will be defined with #3945, so this PR should be blocked on it.

Reply:
No, `IntrusiveDList` was defined with #3907. #3945 is just a minor documentation update.

Reply:
@SolidWallOfCode is correct. It seems there was a bug somewhere in the docs that caused the definition not to be parsed.

It seems now we have a duplicate declaration:
```
/Users/derekd/devel/trafficserver/doc/developer-guide/internal-libraries/intrusive-list.en.rst:46: WARNING: Duplicate declaration.
```

This can be fixed elsewhere.

Reply:
Yes, that's one of the things #3945 fixes.

Reply:
OK.